docs/proposals/go-git-providers.md

```diff
@@ -0,0 +1,474 @@
+# go-git-providers
+
+## Abstract
+
+This proposal aims to create a library with the import path `github.com/fluxcd/go-git-providers`'
+(import name: `gitprovider`), which provides an abstraction layer for talking to Git providers
+like GitHub, GitLab and Bitbucket.
+
+This would become a new repository, specifically targeted at being a general-purpose Git provider
+client for multiple providers and domains.
+
+## Goals
+
+- Support multiple Git provider backends (e.g. GitHub, GitLab, Bitbucket, etc.) using the same interface
+- Support talking to multiple domains at once, including custom domains (e.g. talking to "gitlab.com" and "version.aalto.fi" from the same client)
```

Comment:
What is the motivation for this? It sounds tricky to get right.

Reply:
Indeed, I see no use-case for this.

Reply:
@dinosk expressed some interest in automatically using different backends for parsed clone/repo URLs. This is what the `MultiClient` is doing. I was thinking of it to be dead simple, in that it just checks at the domain of the Git URL, and uses the right client for the right URL. However, `MultiClient` is purely additive, it doesn't affect the rest of the proposal. I thought it might save some typing.

If you strongly feel against it, lmk and I can remove it, at least for the first iteration.

Reply:
Oh, it's a _factory_ for clients. I have no strong objection to code that examines the supplied URL and decides which implementation to use. Might be expedient to leave it for now though.